File: client-py/pending.py
"""Política de retry pura sobre o mapa de pendentes.

pending: dict ref -> {"line", "tick", "retries"}. Nada aqui toca no socket:
on_timeout devolve o que fazer (retry ou drop) e a borda executa.
"""

BUDGET = 3  # tentativas máximas por ref
TIMEOUT = 2  # ticks sem ACK antes de reenviar


def add_pending(pending, ref, line, tick):
    return {**pending, ref: {"line": line, "tick": tick, "retries": 0}}


def on_ack(pending, ref):
    return {k: v for k, v in pending.items() if k != ref}


def on_timeout(pending, now, budget=BUDGET, timeout=TIMEOUT):
    """refs vencidos: retry enquanto couber no orçamento, drop no limite."""
    effects = []
    nxt = {}
    for ref, entry in pending.items():
        if now - entry["tick"] < timeout:
            nxt[ref] = entry
        elif entry["retries"] < budget:
            nxt[ref] = {**entry, "tick": now, "retries": entry["retries"] + 1}
            effects.append({"retry": ref, "line": entry["line"]})
        else:
            effects.append({"drop": ref})
    return nxt, effects

File: client-go/pending.go
package main

// Política de retry pura sobre o mapa de pendentes.
// Nada aqui toca no socket: OnTimeout devolve o que fazer
// (retry ou drop) e a borda executa.

const (
	Budget  = 3 // tentativas máximas por ref
	Timeout = 2 // ticks sem ACK antes de reenviar
)

type PendingEntry struct {
	Line    string
	Tick    int
	Retries int
}

type Effect struct {
	Ref  string
	Line string // preenchida em retry
	Drop bool
}

func AddPending(pending map[string]PendingEntry, ref, line string, tick int) map[string]PendingEntry {
	next := make(map[string]PendingEntry, len(pending)+1)
	for k, v := range pending {
		next[k] = v
	}
	next[ref] = PendingEntry{Line: line, Tick: tick}
	return next
}

func OnAck(pending map[string]PendingEntry, ref string) map[string]PendingEntry {
	next := make(map[string]PendingEntry, len(pending))
	for k, v := range pending {
		if k != ref {
			next[k] = v
		}
	}
	return next
}

// refs vencidos: retry enquanto couber no orçamento, drop no limite.
func OnTimeout(pending map[string]PendingEntry, now int) (map[string]PendingEntry, []Effect) {
	var effects []Effect
	next := make(map[string]PendingEntry, len(pending))
	for ref, entry := range pending {
		switch {
		case now-entry.Tick < Timeout:
			next[ref] = entry
		case entry.Retries < Budget:
			entry.Tick = now
			entry.Retries++
			next[ref] = entry
			effects = append(effects, Effect{Ref: ref, Line: entry.Line})
		default:
			effects = append(effects, Effect{Ref: ref, Drop: true})
		}
	}
	return next, effects
}
